### PR TITLE
immutable-object-cache: fixed crashes on start up

### PR DESCRIPTION
**Pull request by dillaman**
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by trociny** [comment]
jenkins render docs

File: src/tools/immutable_object_cache/ObjectCacheStore.cc [hunk 1]
```diff
@@ -61,16 +61,19 @@ int ObjectCacheStore::init(bool reset) {
...
-      for (auto& p : efs::directory_iterator(m_cache_root_dir)) {
-        efs::remove_all(p.path());
-      }
-    } else {
-      if (!efs::create_directories(m_cache_root_dir, ec)) {
-        lderr(m_cct) << "fail to create cache store dir: " << ec << dendl;
-        return ec.value();
+    try {
+      if (efs::exists(m_cache_root_dir)) {
+        // remove all sub folders
```
Comment (trociny) [inline src/tools/immutable_object_cache/ObjectCacheStore.cc:66]: Due to this I don't like very much that the default `immutable_object_cache_path` is `/tmp`. Consider an admin just learning about this tool and starting with defaults. I think she would not be happy to learn that all subfolders in '/tmp' dir had gone. This is actually what I experienced when first tried it. Luckily it was not from the root so eventually it failed due to a permission error, but who knows what folders it had removed before this... 

Something like `/tmp/ceph/immutable_object_cache` would be much safer.

Comment (dillaman) [reply src/tools/immutable_object_cache/ObjectCacheStore.cc:65]: Good point. Switched to `/tmp/ceph_immutable_object_cache`

File: src/tools/immutable_object_cache/CacheController.cc [hunk 2]
```diff
@@ -64,10 +67,15 @@ void CacheController::handle_signal(int signum) {
   shutdown();
 }
 
-void CacheController::run() {
+int CacheController::run() {
   try {
     std::string controller_path =
       m_cct->_conf.get_val<std::string>("immutable_object_cache_sock");
+    if (controller_path.empty()) {
```
Comment (CongMinYin) [inline src/tools/immutable_object_cache/CacheController.cc:74]: Hi Jason, I want to describe such a scenario: if using default "/var/run/ceph/immutable_object_cache_sock", the root user had started the daemon, then "immutable_object_cache_sock" belongs to root. other users also use the default immutable_object_cache_sock path, they won't have permission . Or specify a immutable_object_cache_sock path that already exists and belongs to root. At this time, there should be a permission problem. Will the code run here and throw an error to tell users? The immutable_object_cache_path is also similar. So, should we suggest users to set a new path that the current user has permission to?

Comment (dillaman) [reply src/tools/immutable_object_cache/CacheController.cc:74]: > So, should we suggest users to set a new path that the current user has permission to?

Yes.

**Comment by trociny** [comment]
jenkins test make check

File: src/tools/immutable_object_cache/CacheController.cc (hunk 2 above)
Comment (kenneth104) [reply src/tools/immutable_object_cache/CacheController.cc:74]: woo,I finally found the problem I encountered

**Pull request merged**